src/EnergyPlus/RoomAirModelManager.cc

```diff
@@ -2693,225 +2714,204 @@ namespace RoomAir {
-        int MaxNums = 0;         // Maximum number of numeric input fields
-        int MaxAlphas = 0;       // Maximum number of alpha input fields
-        int TotalArgs = 0;       // Total number of alpha and numeric arguments(max) for a
-        Array1D_string Alphas;   // Alpha input items for object
-        Array1D<Real64> Numbers; // Numeric input items for object
         bool errorfound;
-
-        NumAlphas = 1;
-        NumNumbers = 1;
-        EquipFind = false;
+        int SupplyNodeNum = 0;
+        int ReturnNodeNum = 0;
 
         SupplyNodeName = "";
 
         if (zoneEquipType == DataZoneEquipment::ZoneEquipType::Invalid) return EquipFind;
 
-        std::string_view equipTypeName = DataZoneEquipment::zoneEquipTypeNamesUC[(int)zoneEquipType];
-        state.dataInputProcessing->inputProcessor->getObjectDefMaxArgs(state, equipTypeName, TotalArgs, NumAlphas, NumNumbers);
-
-        MaxNums = max(MaxNums, NumNumbers);
-        MaxAlphas = max(MaxAlphas, NumAlphas);
-
-        if (MaxNums > NumNumbers) {
-            Numbers.allocate(MaxNums);
-            Numbers = 0.0;
-        } else if (!allocated(Numbers)) {
-            Numbers.allocate(MaxNums);
-        }
-
-        if (MaxAlphas > NumAlphas) {
-            Alphas.allocate(MaxAlphas);
-            Alphas = "";
-        } else if (!allocated(Alphas)) {
-            Alphas.allocate(NumAlphas);
-        }
-
-        for (int I = 1; I <= TotNumEquip; ++I) {
-            state.dataInputProcessing->inputProcessor->getObjectItem(state, equipTypeName, I, Alphas, NumAlphas, Numbers, NumNumbers, Status);
-            if (Util::SameString(Alphas(1), EquipName)) {
-                EquipFind = true;
-                break;
-            }
-        }
-
         switch (zoneEquipType) {
         case DataZoneEquipment::ZoneEquipType::VariableRefrigerantFlowTerminal: { // ZoneHVAC:TerminalUnit : VariableRefrigerantFlow
-            SupplyNodeName = Alphas(4);
-            ReturnNodeName = ""; // Zone return node
+            SupplyNodeNum = HVACVariableRefrigerantFlow::GetVRFTUZoneInletAirNode(state, EquipIndex);
         } break;
         case DataZoneEquipment::ZoneEquipType::EnergyRecoveryVentilator: { // ZoneHVAC : EnergyRecoveryVentilator
-            int nodeNum = GetFanOutletNode(state, "Fan:OnOff", Alphas(4), errorfound);
-            if (errorfound) {
-            }
-            SupplyNodeName = state.dataLoopNodes->NodeID(nodeNum); // ?????
-            ReturnNodeName = "";                                   // Zone exhaust node
+            SupplyNodeNum = GetFanOutletNode(state, "Fan:OnOff", EquipName, errorfound);
         } break;
         case DataZoneEquipment::ZoneEquipType::FourPipeFanCoil: { // ZoneHVAC : FourPipeFanCoil
-            SupplyNodeName = Alphas(6);
-            ReturnNodeName = Alphas(5);
+            SupplyNodeNum = state.dataFanCoilUnits->FanCoil(EquipIndex).AirOutNode;
+            ReturnNodeNum = state.dataFanCoilUnits->FanCoil(EquipIndex).AirInNode;
```

Comment:
Get air inlet and outlet nodes directly from FanCoilUnits based on EquipIndex

Reply:
@Myoldmopar At some point in the past, we frowned on looking directly into other components' data structures, e.g.
` SupplyNodeNum = state.dataFanCoilUnits->FanCoil(EquipIndex).AirOutNode;` 
and preferred using mining functions, e.g. 
` SupplyNodeNum = HVACVariableRefrigerantFlow::GetVRFTUZoneInletAirNode(state, EquipIndex);`

What's our current preference?

Reply:
There are two questions here. One is whether we should use access functions or direct access. There are two benefits to using access functions:
- The ability to control/enforce simultaneous and internally consistent of related member variables.
- The ability to support an external API.
The downside of course is overhead, both in terms of code but also at runtime.

If we want to go with access functions, there are two choices:
- Mining functions that take an object index: overhead is higher, but can double as an API if needed (I think not). We use this pattern fairly extensively.
- Methods: low overhead, but no API support. We hardly use this. 
Unfortunately, there is also a non-choice that we use extensively and that is mining functions that take an object name and have to search for the object on every call. Yikes! We need to sweep this idiom out and the most recent fans PR does that for that module.

I personally favor direct access, but there are some advantages to access methods and if people want to go that round I would not scream too loud.


- Mining functions that take an object name: this is terrible because you are searching for the object again and again, there is too much of it going around. We need to sweep out this pattern.

Reply:
The only reason I see for using mining functions is the call to getInput. If the fan coil has been read in then accessing that data, so the user doesn't have to provide it, seems reasonable (line 2735 above). If the fan coil has not been read in, then a mining function will do the trick. I envision keeping the getIndex for each component type and using direct access for everything else. The problem here is `EquipIndex = state.dataZoneEquip->ZoneEquipList(iZone).EquipIndex(thisZoneEquipNum);` may still be 0 at this point (I didn't trace back to see if all the equipment has been read in) or does not necessarily point to the component index and could instead point to the index from 1-n of the type of zone equipment (e.g. PTAC 1 to 3, or PTHP 1 to 4, etc). So index = 1 would point to UnitarySystem 1 of 7. Since we don't have example files with various mixes of component types, the mistakes don't show up until a user provides a broken input file with a mix of multiple equipment types.

I've tried to explain this concept as best I can in the past.

This is what gets passed back from the simulate function:

        CompIndex = this->m_EquipCompNum;

And this is how that variable gets set (because of zone availability managers):

                        // zone equipment require a 1-n index for access to zone availability managers
                        switch (getPTUnitType) {
                        case 1: // Excuse me?
                            ++numPTAC;
                            thisSys.m_EquipCompNum = numPTAC;
                            break;
                        case 2: // Baking powder?
                            ++numPTHP;
                            thisSys.m_EquipCompNum = numPTHP;
                            break;
                        case 3:
                            ++numPTWSHP;
                            thisSys.m_EquipCompNum = numPTWSHP;
                            break;

Reply:
Out of curiosity, this is what PackagedTerminalHeatPump did prior to #9052:

            auto const SELECT_CASE_var(PTUnitType);
            if (SELECT_CASE_var == PkgTermHPAirToAir_Num) {
                PTUnitNum = CompIndex;
            } else if (SELECT_CASE_var == PkgTermACAirToAir_Num) {
                PTUnitNum = CompIndex + state.dataPTHP->NumPTHP;
            } else if (SELECT_CASE_var == PkgTermHPWaterToAir_Num) {
                PTUnitNum = CompIndex + state.dataPTHP->NumPTHP + state.dataPTHP->NumPTAC;

Reply:
I ran another example file as FanCoilAutoSize.idf to see if which one of getRoomAirModeInput and getFanCoilInput gets first. The getRoomAirModeInput comes first, and getFanCoilInput comes much later. Therefore, the direct access may not work well for every case. In order to be consistent, I use the access functions for every zone equipment type check.